File: Basic Sorting Algorithms/QuickSort.js
function partition(arr, s, e) {
  let i = s - 1;
  let pivot = arr[e];

  for (let j = s; j <= e - 1; j++) {
    if (arr[j] < pivot) {
      [arr[i + 1], arr[j]] = [arr[j], arr[i + 1]];
      i++;
    }
  }

  [arr[i + 1], arr[e]] = [arr[e], arr[i + 1]];
  return i + 1;
}

function quick_sort(arr, s, e) {
  if (s >= e) {
    return;
  }

  let p = partition(arr, s, e);
  quick_sort(arr, s, p - 1);
  quick_sort(arr, p + 1, e);
}

let arr = [-2, 3, 4, -1, 5, -12, 6, 1, 3];
quick_sort(arr, 0, arr.length - 1);

for (let num of arr) {
  console.log(num);
}

File: Recursion And Backtracking/quick_sort.cpp
#include<iostream>
#include<vector>
using namespace std;

int partition(vector<int> & arr, int s, int e) {
    int pivot = arr[e];
    int i = s-1;
    for(int j=s; j<=e-1; j++) {
        if(arr[j] < pivot) {
            swap(arr[i+1], arr[j]);
            i++;
        }
    }
    
    swap(arr[i+1], arr[e]);
    return i+1;
}

void quick_sort(vector<int> & arr, int s, int e) {
    if(s >= e) {
        return;
    }
    
    int p = partition(arr, s, e);
    quick_sort(arr, s, p-1);
    quick_sort(arr, p+1, e);
}


int main() {
    vector<int> arr = {10,5,2,0,7,6,4};
    int s = 0;
    int e = arr.size()-1;
    
    quick_sort(arr, s, e);
    
    for(int num : arr) {
        cout<<num<<", ";
    }
}